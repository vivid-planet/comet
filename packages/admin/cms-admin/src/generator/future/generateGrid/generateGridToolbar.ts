import { camelCase } from "change-case";

import { camelCaseToHumanReadable } from "../utils/camelCaseToHumanReadable";
import { getFormattedMessageNode } from "../utils/intl";

type Options = {
    componentName: string;
    forwardToolbarAction: boolean | undefined;
    hasSearch: boolean;
    hasFilter: boolean;
    excelExport: boolean | undefined;
    allowAdding: boolean;
    instanceGqlType: string;
    gqlType: string;
    newEntryText: string | undefined;
    fragmentName: string;
};

export const generateGridToolbar = ({
    componentName,
    forwardToolbarAction,
    hasSearch,
    hasFilter,
    excelExport,
    allowAdding,
    instanceGqlType,
    gqlType,
    newEntryText,
    fragmentName,
}: Options) => {
    return `${renderToolbarProps(componentName, !!forwardToolbarAction, !!excelExport)}
    function ${componentName}(${getGridToolbarProps(componentName, !!forwardToolbarAction, !!excelExport)}) {
        return (
            <DataGridToolbar>
                ${hasSearch ? searchItem : ""}
                ${hasFilter ? filterItem : ""}
                <ToolbarFillSpace />
                ${renderToolbarActions({
                    forwardToolbarAction,
                    addItemText: getFormattedMessageNode(
                        `${instanceGqlType}.${camelCase(fragmentName)}.newEntry`,
                        newEntryText ?? `New ${camelCaseToHumanReadable(gqlType)}`,
                    ),
                    excelExport,
                    allowAdding,
                })}
            </DataGridToolbar>
        );
    }`.replace(/^\s+\n/gm, "");
};

const getGridToolbarProps = (componentName: string, toolbarAction: boolean, exportApi: boolean) => {
    const props: Array<{
        destructured: string;
        typeDefinition: string;
    }> = [];

    if (toolbarAction) {
        props.push({
            destructured: "toolbarAction",
            typeDefinition: "toolbarAction?: ReactNode",
        });
    }

    if (exportApi) {
        props.push({
            destructured: "exportApi",
            typeDefinition: "exportApi: ExportApi",
        });
    }

    if (!props.length) {
        return "";
    }

    return `{
        ${props.map((prop) => `${prop.destructured}`).join(",")}
    }: ${componentName}ToolbarProps`;
};

const searchItem = `<ToolbarItem>
    <GridToolbarQuickFilter />
</ToolbarItem>`;

const filterItem = `<ToolbarItem>
    <GridFilterButton />
</ToolbarItem>`;

type RenderToolbarActionsOptions = {
    forwardToolbarAction: boolean | undefined;
    addItemText: string;
    excelExport: boolean | undefined;
    allowAdding: boolean | undefined;
};

const renderToolbarActions = ({ forwardToolbarAction, addItemText, excelExport, allowAdding }: RenderToolbarActionsOptions) => {
    const showMoreActionsMenu = excelExport;

    if (!showMoreActionsMenu && !allowAdding) {
        return "";
    }

    const moreActionsMenu = `<CrudMoreActionsMenu
        slotProps={{
            button: {
                responsive: true
            }
        }}
        overallActions={[
            ${
                excelExport
                    ? `{
                        label: <FormattedMessage {...messages.downloadAsExcel} />,
                        icon: exportApi.loading ? <CircularProgress size={20} /> : <Excel />,
                        onClick: () => exportApi.exportGrid(),
                        disabled: exportApi.loading,
                    }`
                    : ""
            }
        ]}
    />`;

    const defaultAddItemButton = `<Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        ${addItemText}
    </Button>`;

    const addAction = forwardToolbarAction ? "{toolbarAction}" : defaultAddItemButton;

    return `<ToolbarActions>
        ${showMoreActionsMenu ? moreActionsMenu : ""}
        ${allowAdding ? addAction : ""}
    </ToolbarActions>`;
};

const renderToolbarProps = (componentName: string, forwardToolbarAction: boolean | undefined, exportApi: boolean) => {
    if (forwardToolbarAction || exportApi) {
        return `interface ${componentName}ToolbarProps extends GridToolbarProps {
         ${forwardToolbarAction && "toolbarAction: ReactNode;"}
         ${exportApi ? "exportApi: ExportApi;" : ""}
}`;
    }
    return "";
};
