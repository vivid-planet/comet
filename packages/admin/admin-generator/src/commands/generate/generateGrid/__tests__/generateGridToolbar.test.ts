import { generateGridToolbar } from "../generateGridToolbar";

describe("generateGridToolbar", () => {
    it("generates toolbar with all features enabled", () => {
        const output = generateGridToolbar({
            componentName: "BooksGridToolbar",
            forwardToolbarAction: true,
            hasSearch: true,
            hasFilter: true,
            excelExport: true,
            allowAdding: true,
            instanceGqlType: "booksMessages",
            gqlType: "Book",
            newEntryText: undefined,
            fragmentName: "bookFragment",
        });

        expect(output).toMatchSnapshot();
    });

    it("generates toolbar with minimal features", () => {
        const output = generateGridToolbar({
            componentName: "BooksGridToolbar",
            forwardToolbarAction: false,
            hasSearch: false,
            hasFilter: false,
            excelExport: false,
            allowAdding: true,
            instanceGqlType: "booksMessages",
            gqlType: "Book",
            newEntryText: undefined,
            fragmentName: "bookFragment",
        });

        expect(output).toMatchSnapshot();
    });

    it("generates toolbar with custom newEntryText", () => {
        const output = generateGridToolbar({
            componentName: "BooksGridToolbar",
            forwardToolbarAction: false,
            hasSearch: false,
            hasFilter: false,
            excelExport: false,
            allowAdding: true,
            instanceGqlType: "booksMessages",
            gqlType: "Book",
            newEntryText: "Create new book",
            fragmentName: "bookFragment",
        });

        expect(output).toMatchSnapshot();
    });

    it("generates empty toolbar when no features are enabled", () => {
        const output = generateGridToolbar({
            componentName: "BooksGridToolbar",
            forwardToolbarAction: false,
            hasSearch: false,
            hasFilter: false,
            excelExport: false,
            allowAdding: false,
            instanceGqlType: "booksMessages",
            gqlType: "Book",
            newEntryText: undefined,
            fragmentName: "bookFragment",
        });

        expect(output).toMatchSnapshot();
    });

    it("generates valid Toolbar props for only excelExport enabled", () => {
        const output = generateGridToolbar({
            componentName: "BooksGridToolbar",
            forwardToolbarAction: false,
            hasSearch: false,
            hasFilter: false,
            excelExport: true,
            allowAdding: false,
            instanceGqlType: "booksMessages",
            gqlType: "Book",
            newEntryText: undefined,
            fragmentName: "bookFragment",
        });

        expect(output).toMatchSnapshot();
    });
});
