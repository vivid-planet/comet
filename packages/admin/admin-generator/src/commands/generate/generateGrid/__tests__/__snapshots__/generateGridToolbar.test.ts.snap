// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateGridToolbar generates empty toolbar when no features are enabled 1`] = `
"
    function BooksGridToolbar() {
        return (
            <DataGridToolbar>
                <FillSpace />
            </DataGridToolbar>
        );
    }"
`;

exports[`generateGridToolbar generates toolbar with all features enabled 1`] = `
"interface BooksGridToolbarToolbarProps extends GridToolbarProps {
         toolbarAction: ReactNode;
         exportApi: ExportApi;
}
    function BooksGridToolbar({
        toolbarAction,exportApi
    }: BooksGridToolbarToolbarProps) {
        return (
            <DataGridToolbar>
                <GridToolbarQuickFilter />
                <GridFilterButton />
                <FillSpace />
        <CrudMoreActionsMenu
        slotProps={{
            button: {
                responsive: true
            }
        }}
        overallActions={[
            {
                        label: <FormattedMessage {...messages.downloadAsExcel} />,
                        icon: exportApi.loading ? <CircularProgress size={20} /> : <ExcelIcon />,
                        onClick: () => exportApi.exportGrid(),
                        disabled: exportApi.loading,
                    }
        ]}
    />
        {toolbarAction}
            </DataGridToolbar>
        );
    }"
`;

exports[`generateGridToolbar generates toolbar with custom newEntryText 1`] = `
"
    function BooksGridToolbar() {
        return (
            <DataGridToolbar>
                <FillSpace />
        <Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        <FormattedMessage id="booksMessages.bookFragment.newEntry" defaultMessage={\`Create new book\`}  />
    </Button>
            </DataGridToolbar>
        );
    }"
`;

exports[`generateGridToolbar generates toolbar with minimal features 1`] = `
"
    function BooksGridToolbar() {
        return (
            <DataGridToolbar>
                <FillSpace />
        <Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        <FormattedMessage id="booksMessages.bookFragment.newEntry" defaultMessage={\`New Book\`}  />
    </Button>
            </DataGridToolbar>
        );
    }"
`;

exports[`generateGridToolbar generates valid Toolbar props for only excelExport enabled 1`] = `
"interface BooksGridToolbarToolbarProps extends GridToolbarProps {
         exportApi: ExportApi;
}
    function BooksGridToolbar({
        exportApi
    }: BooksGridToolbarToolbarProps) {
        return (
            <DataGridToolbar>
                <FillSpace />
        <CrudMoreActionsMenu
        slotProps={{
            button: {
                responsive: true
            }
        }}
        overallActions={[
            {
                        label: <FormattedMessage {...messages.downloadAsExcel} />,
                        icon: exportApi.loading ? <CircularProgress size={20} /> : <ExcelIcon />,
                        onClick: () => exportApi.exportGrid(),
                        disabled: exportApi.loading,
                    }
        ]}
    />
            </DataGridToolbar>
        );
    }"
`;
